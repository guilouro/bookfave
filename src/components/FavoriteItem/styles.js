import styled from 'styled-components'
import { CleanButton, colors } from '../../utils.styles'

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 15px 0;
  &:not(:last-child) {
    border-bottom: 1px solid #e7e8eb;
  }
`

export const Info = styled.div``
export const Tags = styled.div``

export const Title = styled.h2`
  font-size: 16px;
  font-weight: 300;
  margin: 0;
`

export const Link = styled.a`
  font-size: 12px;
  color: ${colors.link};
  text-decoration: none;
  margin: 10px 0;
  display: table;
`

export const Delete = styled(CleanButton)`
  display: flex;
  align-items: center;
  text-transform: uppercase;
  color: ${colors.secondary};
  span {
    margin-left: 5px;
  }
`
