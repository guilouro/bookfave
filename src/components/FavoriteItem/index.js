import React from 'react'
import PropTypes from 'prop-types'
import * as S from './styles'
import Tag from '../Tag'
import Icon from '../Icons'

const formatLink = link => {
  const isValid = /^(http|\/\/)/i.test(link)
  return isValid ? link : `//${link}`
}

const FavoriteItem = ({
  title,
  link,
  tags,
  onRemove,
  onFilterTag,
  onRemoveTag
}) => (
  <S.Container>
    <S.Info>
      <S.Title>{title}</S.Title>
      <S.Link href={formatLink(link)} target="_blank">
        {link}
      </S.Link>

      <S.Tags>
        {tags.map(tag => (
          <Tag
            key={tag}
            name={tag}
            onClick={() => onFilterTag(tag)}
            onClose={() => onRemoveTag(tag)}
          />
        ))}
      </S.Tags>
    </S.Info>
    <S.Delete onClick={onRemove}>
      <Icon name="Trash" size="16px" />
      <span>Delete</span>
    </S.Delete>
  </S.Container>
)

FavoriteItem.propTypes = {
  onFilterTag: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  onRemoveTag: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  tags: PropTypes.array
}

FavoriteItem.defaultProps = {
  tags: []
}

export default FavoriteItem
