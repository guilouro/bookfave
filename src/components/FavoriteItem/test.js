import React from 'react'
import FavoriteItem from '.'
import { render, fireEvent, cleanup } from 'react-testing-library'

const props = {
  id: '1',
  title: 'Crédito',
  link: 'www.bebluecredito.com.br',
  tags: ['crédito'],
  onRemove: jest.fn(),
  onRemoveTag: jest.fn(),
  onFilterTag: jest.fn()
}

describe('<FavoriteItem />', () => {
  afterEach(cleanup)

  it('Should render correctly', () => {
    const { container, queryByText } = render(<FavoriteItem {...props} />)

    expect(queryByText(props.title)).toBeTruthy()
    expect(queryByText(props.link)).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('Should call onRemove', () => {
    const onRemove = jest.fn()
    const { getByText } = render(
      <FavoriteItem {...props} onRemove={onRemove} />
    )

    fireEvent.click(getByText(/Delete/g))
    expect(onRemove).toBeCalled()
  })

  it('Should call onRemoveTag', () => {
    const onRemoveTag = jest.fn()
    const { getByText } = render(
      <FavoriteItem {...props} onRemoveTag={onRemoveTag} />
    )

    fireEvent.click(getByText(/✕/g))
    expect(onRemoveTag).toBeCalledWith(props.tags[0])
  })

  it('Should call onFilterTag', () => {
    const onFilterTag = jest.fn()
    const { getByText } = render(
      <FavoriteItem {...props} onFilterTag={onFilterTag} />
    )

    fireEvent.click(getByText(props.tags[0]))
    expect(onFilterTag).toBeCalledWith(props.tags[0])
  })

  it('Should resolve link without http', () => {
    const { container } = render(
      <FavoriteItem {...props} link="www.google.com" />
    )
    expect(container.querySelector('a').getAttribute('href')).toBe(
      '//www.google.com'
    )
  })

  it('Should resolve link without http and www', () => {
    const { container } = render(<FavoriteItem {...props} link="google.com" />)
    expect(container.querySelector('a').getAttribute('href')).toBe(
      '//google.com'
    )
  })

  it('Should resolve link http', () => {
    const { container } = render(
      <FavoriteItem {...props} link="http://google.com" />
    )
    expect(container.querySelector('a').getAttribute('href')).toBe(
      'http://google.com'
    )
  })
})
