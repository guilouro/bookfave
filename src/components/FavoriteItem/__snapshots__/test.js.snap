// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteItem /> Should render correctly 1`] = `
<div>
  <div
    class="sc-bdVaJa djegqS"
  >
    <div
      class="sc-bwzfXH uTxCW"
    >
      <h2
        class="sc-bxivhb hBvQiR"
      >
        Crédito
      </h2>
      <a
        class="sc-ifAKCX dVgucK"
        href="//www.bebluecredito.com.br"
        target="_blank"
      >
        www.bebluecredito.com.br
      </a>
      <div
        class="sc-htpNat KtFsv"
      >
        <span
          class="sc-gzVnrw bWPtIe"
        >
          <span>
            crédito
          </span>
          <button
            class="sc-htoDjs sc-dnqmqq haMeb"
          >
            ✕
          </button>
        </span>
      </div>
    </div>
    <button
      class="sc-EHOje sc-bZQynM bzKIkI"
    >
      <svg
        height="16px"
        viewBox="0 0 16 18"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 15c0 .551-.448 1-1 1H5c-.552 0-1-.449-1-1V6h8v9zM8 2c1.103 0 2 .897 2 2H6c0-1.103.897-2 2-2zm7 2h-3c0-2.206-1.794-4-4-4S4 1.794 4 4H1a1 1 0 1 0 0 2h1v9c0 1.654 1.346 3 3 3h6c1.654 0 3-1.346 3-3V6h1a1 1 0 1 0 0-2z"
          fill="currentColor"
        />
      </svg>
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;
